### ISSUE TITLE
Pure ruby replacement for `git ls-files` in the bundle gem generated gemspec

### ISSUE THREAD

**Issue by robotdana** [rubygems/rubygems#4047]
This can make git an unexpected & barely-necessary dependency of many gems.

e.g. https://github.com/rubygems/bundler/issues/5810

I have two gems which i'd be happy to copy into rubygems/bundler somehow to avoid this dependency on git
- https://github.com/robotdana/git_ls (parse the binary .git/index file in pure ruby)
- https://github.com/robotdana/fast_ignore (reads all .gitignore rules in pure ruby and walks the file system checking against them, useful when .git/index doesn't exist for some reason. it's also pretty quick)

Is this something the bundler team would be at all interested in me doing?

**Comment by simi** [comment]
@robotdana It would be great to find alternative having no external dependency. But how should that work in your case? If I understand this well, your idea is to vendor those gems into rubygems and use them in default gemspecs. What about older rubygems versions? AFAIK it will make that new gemspec incompatible with old versions.

**Comment by deivid-rodriguez** [comment]
Hi @robotdana! Thanks for bringing this up.

Yes, I would definitely be interested. I guess the idea would be to vendor this code inside rubygems and the use it in the generated gemspec if available, right?

Something I was thinking though after I saw https://github.com/rubygems/rubygems/pull/4033 is that maybe "respecting git" is not a good idea after all. If I got the history of this issue correctly, we introduced `git ls-files` in the generated gemspec to avoid newbie users from pushing heavy gems full of "garbage files" like generated files. However, `git ls-files` does not prevent "garbage" from being included on a gem as that pull request shows. In other words, you might want some files to be source controlled, but not included with your gem. Conversely, you might want some generated files to be actually included. For example, in the case of gems including precompiled extensions.

So, I'm leaning towards providing a good explicit enough default, like `Dir["lib/**/*.rb"]`, and also `Dir["ext/**/*.c"]` if `bundle gem` is run with the `--ext` flag.

Thoughts?

**Comment by robotdana** [comment]
@simi, regarding the backward compatibility thing i was imagining at least temporarily some kind of `` if defined?(new thing) new thing; else `git ls-files` ``

@deivid-rodriguez I think a sensible simple explicit default would be ideal, and is what i generally do, even on the only gems in the world that can theoretically FastIgnore, i didn't bother suggesting it because it felt like the team was set on respecting git

The other advantage of `Dir something`, is its (hopefully) simpler ruby and therefore less intimidating to edit from the default, the current generated `files=` is quite convoluted with having to chdir and split on nulls and all the regexes

**Comment by simi** [comment]
Just to start from the scratch, what's the problem you're trying to solve @robotdana? If the problem is missing `git` on system, what about to check for git presence during `bundle gem` and provide gem skeleton working on the current system.

I'm also aware of some problems with `git` dependency for Debian (and maybe Fedora) automated package builds.

**Comment by deivid-rodriguez** [comment]
> @deivid-rodriguez I think a sensible simple explicit default would be ideal, and is what i generally do, even on the only gems in the world that can theoretically FastIgnore, i didn't bother suggesting it because it felt like the team was set on respecting git

Yes, you're correct that has been the team's take on this so far. However, I think we could maybe revisit it now. It seems to me that `Dir["lib/**/*.rb"]` is a sensible default since it prevents non source controlled stuff in `lib` from being pushed (unless it's a generated gitignored ruby file, which is something I believe I've never seen).

**Comment by robotdana** [comment]
@simi the debian package builds is the issue i'm aware of, as someone on that team brought it up to me to see if my gems could be useful to solve that problem. it didn't seem possible just by making my gem a requirement in the gemspec somehow (possibly is, but it seems like it would need to happen too early in the process) here i am

**Comment by simi** [comment]
@deivid-rodriguez regarding `Dir[]`, I'm afraid that would result into committing non-complete code to rubygems.org without noticing since majority of tools around common gem infrastructure (for example CI) are related to source code which is usually versioned by `git`. In my :eyes: it is not common to build gem on CI and test it works.

**Comment by robotdana** [comment]
as an aside it'd be neat if the generated spec_helper built the gem and required that rather than just requiring the source

**Comment by deivid-rodriguez** [comment]
@simi I'm sure there's some tiny percentage of gems that need some files to work outside of `Dir["lib/**/*.rb"]` and `Dir["ext/**/*.c"]`. It seems like a really small minority, but I guess this could cause some issues if a popular gem using the new default adds a set of `yml` files for example, and pushes a release without them. I guess we could make the way of generating the file list configurable with a setting like `bundle config gem.file_list <git|glob|explicit>` or something like that.

In any case, vendoring some ruby code that implements `git ls-files` to replace the current template seems like a good idea to me.

**Comment by deivid-rodriguez** [comment]
In conclusion, I think the overall team sentiment is going to be that the most conservative option for now is to keep respecting `git` but I don't see any downsides of "respecting git in ruby", except for vendoring some external code, which I don't mind.

**Comment by robotdana** [comment]
cool i'll have a go the next weekend or so preparing these gems to be used by more than just me (i want to tidy up the api of FastIgnore a bit, and have it use GitLS by default if the .git/index is there) and then have a go at vendoring the code and updating the template

Thanks for your time

**Comment by deivid-rodriguez** [comment]
Let's wait some days for more opinions just in case but I think the conclusion from https://github.com/rubygems/bundler/issues/5810 is that we would accept what you are proposing is someone implemented it, so you should be good to move this forward.

Thanks for your work too! :heart:

**Comment by simi** [comment]
So would be the transitional period in generated gemspec using following snippet?

```ruby
if rubygems_version >= 'x.y.z'
  # use vendored git logic to find files
else
  # use system git
end
```

My opinion in here is current approach works well as is. If we just swap implementation to use vendored git code instead of system git, users should not notice (if implemented well) and it will be just implementation detail I'm ok with.

I'm against moving to some default `Dir[]` since there's risk to break current user workflow (at least AFAIK). I don't see any reason to change something that works (AFAIK) well for users.

Regarding packaging issues, @robotdana are there any details you can link me to to explain what's the problem adding `git` as a build dependency? I think I have seen some explanation before, but that was probably Fedora related. Anyway Debian/Fedora packages of rubygems are really minor usecase AFAIK, majority of rubygems users are using custom installation. @deivid-rodriguez do we have any insight (stats) into this as well?

**Comment by deivid-rodriguez** [comment]
> My opinion in here is current approach works well as is. If we just swap implementation to use vendored git code instead of system git, users should not notice (if implemented well) and it will be just implementation detail I'm ok with.

Sure, let's do this fully transparent for users.

> I'm against moving to some default Dir[] since there's risk to break current user workflow (at least AFAIK). I don't see any reason to change something that works (AFAIK) well for users.

I'd be in favour of switching but I think it should be a separate discussion from this issue, since this one should be 100% transparent for everyone. So I retire my `Dir[]` proposal so that we can focus on this issue.

> Regarding packaging issues, @robotdana are there any details you can link me to to explain what's the problem adding git as a build dependency? I think I have seen some explanation before, but that was probably Fedora related. Anyway Debian/Fedora packages of rubygems are really minor usecase AFAIK, majority of rubygems users are using custom installation. @deivid-rodriguez do we have any insight (stats) into this as well?

No, just that I do care about making the experience better for ruby packagers, and that it feels nice to not have to shell out to an external tool every time a gemspec is evaluated. I've also seen this cause issues for final `Gemfile` users inside `docker` images not including `git`, when running running `bundle install` of a pre populated gem cache of a `Gemfile` using `:git` sources. Technically this shouldn't need `git` if everything is properly cloned beforehand, but it leads to confusing "missing git" errors being printed.

**Comment by utkarsh2102** [comment]
Oh hey everyone 👋🏻 

Thankfully, somehow, I stumbled upon this issue and I'm glad I did! Thanks a lot, @robotdana, for raising this up finally! 😄 
FWIW, there's [RuboCop::Packaging](https://docs.rubocop.org/rubocop-packaging/)'s documentation [explaining the problems of using `git`](https://docs.rubocop.org/rubocop-packaging/cops_packaging.html#gemspec-git-rationale).

Interestingly, it has been well received so far and people seem to have been liking using `Dir`/`Dir.glob`. There's also [rake-manifest](https://github.com/mvz/rake-manifest) gem by @mvz which is also what some people prefer!

In my opinion, it's definitely a very good idea to move away from `git` altogether and mostly every upstream is in favor of that! What's stopping them is the fear of pushing a gem with some useless file that should have been ignored by git (using the `.gitignore` file!).

I loved @robotdana's work on `fast_ignore` and `git_ls`, I think that is definitely a potential solution for getting rid of `git` altogether!
(also, she was the one who helped in setting up the inital scaffold of the `GemspecGit` cop in `RuboCop::Packaging`, so thanks again for that, too! ❤️)

One last thing I want to add is `Rake::FileList`. It's yet another potential solution for this problem. The only problem pursuing this is that using something like this `Rake::FileList["lib/**/*"].exclude(*File.read('.gitignore').split)` works nice but only unfortunately that `FileList` cannot yet parse all the rules of `.gitignore`, which is what `fast_ignore` does beautifully!
So it might also be worth working on Rake's `FileList` method.

Either way, I am happy that this issue is being discussed again and hopefully we'll get away from using git for once and all! \o/
Thank you, everyone, for your work on this! ❤️

**Comment by rubyFeedback** [comment]
My personal opinion is quite simple: ideally we'd never have to use anything but ruby, so
all external calls to git can be problematic. So because of this I agree with the threadstarter.

I think a bigger problem is how git sometimes does unexpected things, in particular when
these may also fail. I give you a non-git example: meson tends to do git checkouts if
a dependency is missing. Since I hate that (I track all dependencies via a ruby gem
already as-is), it took me a while to figure out that --wrap-mode=nodownload
can disable this. But I did not know that for a few months. So that always adds more
time one has to spend in order to workaround for problems induced by other programs,
and git is quite complex and also complicated in all its possibilities.

**Comment by rubyFeedback** [comment]
As for Rake::FileList - I'd actually rather prefer to see the Rake DSL gone and
instead, pure ruby is used rather than such arbitrary ad-hoc solutions such as
FileList. Yes, I know, this may be a minority opinion because lots of people
like rake. Personally I don't like it at all. It always induced more complexity
to my projects when I based them on it. From what I see, rake is mostly just
a "better" make, but make in itself is also quite horrible (I am also aware
that this is not a popular opinion either, but I hate projects that are only
make-based. In my opinion, these are ALWAYS worse than e. g. meson-based
or configure-based projects ... and configure-based projects have another
problem, e. g. the horrible macro-language they use. Actually I much prefer
meson/ninja these days in general. Just sad that not ruby is used ...)

**Comment by mvz** [comment]
A more practical problem with `Rake::FileList` is that it is not always available at the point where a gemspec is first interpreted, so the simple `bundle install; bundle exec rake` instruction fails in the first step.

**Comment by duckinator** [comment]
Any solution we use would need to be included with RubyGems, and we would need an if/else check in the generated gemspecs to keep it from breaking on older RubyGems.

Looking at @robotdana's code, it all looks pretty good, and I would definitely be interested in it being added to Bundler so we can lose the git dependency! :+1: 

Also, @simi suggested using a version check (`if rubygems_version >= 'x.y.z'`), but I think having a function that uses the vendored code (something like `Gem.git_ls_files`/`Bundler.git_ls_files`) and using a feature check (e.g. `if Gem.respond_to?(:git_ls_files)`) would be more robust. It provides a defined API, which gives us the freedom to change the implementation details behind the scenes. Aside from that, I agree with what @simi said.

I'm not sure what the best location or name for that wrapper function would be, though, so it may make sense to poke someone else about that &mdash; @simi, @deivid-rodriguez any thoughts?

**Comment by mvz** [comment]
At the risk of tooting my own horn too much, my rake-manifest gem works with a gemspec that does not depend on rake-manifest. No new code is needed in RubyGems since it just uses Ruby's existing file reading features.

**Comment by deivid-rodriguez** [comment]
Never responded here @duckinator, apologies. Your approach looks good. I think `Gem.git_ls_files` or `Gem::Util.git_ls_files` are ok.

**Comment by hadmut** [comment]
Do I understand this correctly?

There's a broken-by-design-bug, and even after five years you still resist to fix it? ( #8959 )

Then it is probably time to leave ruby ...

**Comment by deivid-rodriguez** [comment]
No, you got that wrong. This is an open issue, and if you read though it, we agreed that it'd be better to remove the dependency on git.

The decision of respecting source control tracked files was intentional to avoid accidental inclusion of unwanted files inside gems. The team decided to keep that decision in place, but use a pure-ruby implementation to build the file list.

And that's the current state of things.

**Comment by hsbt** [comment]
From my experience, `Dir.glob` may include unnecessary files such as temporary files in gem package, so I will continue to use `git` as a template.

`gemspec` is just a template, and you can freely rewrite it if you don't want to use external commands like `git`.
